File: packages/kits/PersistenceKit/Sources/PersistenceKit/RowStore.swift
// RowStore.swift
//
// Typed row I/O protocol.

import Foundation
import SubstrateTypes

public typealias RowKey = UUID

public struct StorageRow: Sendable {
    public let values: [String: TypedValue]

    public init(values: [String: TypedValue]) {
        self.values = values
    }

    public subscript(column: String) -> TypedValue? {
        values[column]
    }
}

public struct RowHandle: Sendable, Hashable {
    public let table: String
    public let key: RowKey

    public init(table: String, key: RowKey) {
        self.table = table
        self.key = key
    }
}

public protocol RowStore: Sendable {
    func insert(table: String, values: [String: TypedValue]) async throws -> RowHandle
    @discardableResult
    func upsert(table: String, values: [String: TypedValue], conflictColumns: [String]) async throws -> RowHandle
    @discardableResult
    func update(table: String, values: [String: TypedValue], where: StoragePredicate) async throws -> Int
    func delete(table: String, where: StoragePredicate) async throws -> Int
    func query(
        table: String,
        where predicate: StoragePredicate?,
        orderBy: [OrderClause],
        limit: Int?,
        offset: Int?
    ) async throws -> [StorageRow]
    func count(table: String, where predicate: StoragePredicate?) async throws -> Int

    /// Corpus scan that skips rows with a corrupt stored value rather than
    /// failing the entire scan.
    ///
    /// ## When to use
    ///
    /// Use this method for **best-effort corpus scans** (e.g. `allDrawers`,
    /// `drawersIn(wing:)`) where a single corrupt row must not brick the entire
    /// estate. For **point lookups** (single-row fetches by primary key) use
    /// `query(...)`, which is strict: a corrupt value in a point-lookup row is
    /// an unambiguous data-integrity failure and the caller must know about it.
    ///
    /// ## Contract
    ///
    /// Returns `(cleanRows, skippedCount)`. Rows that decode without error
    /// appear in `cleanRows`. Rows that fail with `StorageError.corruptStoredValue`
    /// are counted in `skippedCount` and logged via OSLog. Any other error
    /// (engine failure, connectivity, locking) is re-thrown — those are systemic
    /// failures, not data problems.
    ///
    /// ## Default implementation
    ///
    /// Calls `query(...)` and promotes a top-level `corruptStoredValue` error to
    /// `([], 1)`. Backends that iterate at the cursor level (SQLiteStorage)
    /// override this to skip individual corrupt rows without aborting the scan.
    func querySkipCorrupt(
        table: String,
        where predicate: StoragePredicate?,
        orderBy: [OrderClause],
        limit: Int?,
        offset: Int?,
        columns: [String]?
    ) async throws -> (rows: [StorageRow], skipped: Int)

    /// Column-projecting query: like `query(...)` but reads ONLY the named
    /// `columns` from the row, leaving every unnamed column absent from the
    /// returned `StorageRow`. A `nil` projection is a full read (every column).
    ///
    /// This is the no-blob read path: passing the structured/bitmap columns
    /// without the content column means the content blob is never transferred
    /// out of storage — the dense-first candidate-pool load. The returned rows
    /// carry only the projected keys, so a consumer that decodes an absent
    /// column reads the type's empty/default value.
    ///
    /// The protocol-extension default below ignores `columns` and performs the
    /// existing full `query(...)`. Backends on the recall hot path (SQLite,
    /// InMemory) override it to genuinely project; other backends inherit the
    /// full-read default, which is always correct (a superset of the requested
    /// columns) and simply does not realize the no-blob saving.
    func query(
        table: String,
        where predicate: StoragePredicate?,
        orderBy: [OrderClause],
        limit: Int?,
        offset: Int?,
        columns: [String]?
    ) async throws -> [StorageRow]

    // MARK: - Transaction boundary (GLK_BATCH1)

    /// Open a write transaction on the backing store.
    ///
    /// Declared as a protocol requirement (not just a protocol-extension
    /// default) so that dynamic dispatch through `any RowStore` existentials
    /// reaches the concrete type's override. SQLiteRowStore overrides with
    /// `BEGIN IMMEDIATE`. CachingRowStore overrides with explicit delegation
    /// to its backing store. All other conformers inherit the no-op default
    /// provided in the protocol extension.
    func beginTransaction() async throws

    /// Commit the current transaction.
    ///
    /// Protocol requirement for the same dynamic-dispatch reason as
    /// `beginTransaction`.
    func commitTransaction() async throws

    /// Roll back the current transaction, discarding all changes since
    /// the last `beginTransaction` call.
    ///
    /// Protocol requirement for the same dynamic-dispatch reason as
    /// `beginTransaction`.
    func rollbackTransaction() async throws
}

public extension RowStore {
    func query(table: String, where predicate: StoragePredicate? = nil) async throws -> [StorageRow] {
        try await query(table: table, where: predicate, orderBy: [], limit: nil, offset: nil)
    }

    /// Default projection: ignore `columns` and perform a full read. Correct
    /// for every backend (the full row is a superset of any projection); only
    /// the overriding backends realize the no-blob transfer saving.
    func query(
        table: String,
        where predicate: StoragePredicate?,
        orderBy: [OrderClause],
        limit: Int?,
        offset: Int?,
        columns: [String]?
    ) async throws -> [StorageRow] {
        try await query(
            table: table, where: predicate,
            orderBy: orderBy, limit: limit, offset: offset)
    }

    /// Default skip-corrupt implementation: calls `query(...)` with the given
    /// projection and promotes a top-level `StorageError.corruptStoredValue`
    /// error to `([], 1)`. Backends that iterate at cursor level override this
    /// for correct per-row skip-and-log behaviour.
    func querySkipCorrupt(
        table: String,
        where predicate: StoragePredicate?,
        orderBy: [OrderClause],
        limit: Int?,
        offset: Int?,
        columns: [String]?
    ) async throws -> (rows: [StorageRow], skipped: Int) {
        do {
            let rows = try await query(
                table: table, where: predicate,
                orderBy: orderBy, limit: limit, offset: offset, columns: columns)
            return (rows, 0)
        } catch StorageError.corruptStoredValue {
            // Promote: the whole-query abort becomes a skip-and-empty-result.
            // Backends that iterate at cursor level override this for per-row
            // skip-and-log behaviour (SQLiteRowStore overrides this).
            return ([], 1)
        }
    }

    // MARK: - As-of temporal query (ADR-017 §15)

    /// Temporal query: returns rows visible at the given `AsOfCoordinate`.
    ///
    /// - `.present` or `nil`: delegates to the standard `query(...)` — the
    ///   current live state.
    /// - `.asOf(hlc)`: returns rows whose HLC validity range includes the
    ///   given HLC. **Currently gated off** — returns
    ///   `StorageError.featureGated` until NT-L4 (lineage-wide expunge)
    ///   and NT-P3 (erasure overlay) have both merged.
    ///
    /// The filter logic (when ungated): a row with `created_hlc` and optional
    /// `tombstoned_hlc` is visible at HLC T when
    /// `created_hlc <= T AND (tombstoned_hlc IS NULL OR tombstoned_hlc > T)`.
    ///
    /// Default implementation handles the gate. Backends override once the
    /// gate is lifted to push the temporal filter into the engine.
    func query(
        table: String,
        where predicate: StoragePredicate?,
        orderBy: [OrderClause],
        limit: Int?,
        offset: Int?,
        asOf: AsOfCoordinate?
    ) async throws -> [StorageRow] {
        switch asOf {
        case nil, .present:
            return try await query(
                table: table, where: predicate,
                orderBy: orderBy, limit: limit, offset: offset)
        case .asOf:
            throw StorageError.featureGated(feature: "asOfQuery")
        }
    }

    /// Temporal projected query: as-of variant of the column-projecting
    /// query. Same gating behavior as `query(..., asOf:)`.
    func query(
        table: String,
        where predicate: StoragePredicate?,
        orderBy: [OrderClause],
        limit: Int?,
        offset: Int?,
        columns: [String]?,
        asOf: AsOfCoordinate?
    ) async throws -> [StorageRow] {
        switch asOf {
        case nil, .present:
            return try await query(
                table: table, where: predicate,
                orderBy: orderBy, limit: limit, offset: offset,
                columns: columns)
        case .asOf:
            throw StorageError.featureGated(feature: "asOfQuery")
        }
    }

    /// Temporal skip-corrupt query: as-of variant of `querySkipCorrupt`.
    /// Same gating behavior as `query(..., asOf:)`.
    func querySkipCorrupt(
        table: String,
        where predicate: StoragePredicate?,
        orderBy: [OrderClause],
        limit: Int?,
        offset: Int?,
        columns: [String]?,
        asOf: AsOfCoordinate?
    ) async throws -> (rows: [StorageRow], skipped: Int) {
        switch asOf {
        case nil, .present:
            return try await querySkipCorrupt(
                table: table, where: predicate,
                orderBy: orderBy, limit: limit, offset: offset,
                columns: columns)
        case .asOf:
            throw StorageError.featureGated(feature: "asOfQuery")
        }
    }

    // MARK: - Transaction boundary defaults (GLK_BATCH1)

    /// No-op default for `beginTransaction`. Correct for any backend that has
    /// no serializable multi-statement transaction concept (in-memory, hashing).
    func beginTransaction() async throws {}

    /// No-op default for `commitTransaction`.
    func commitTransaction() async throws {}

    /// No-op default for `rollbackTransaction`.
    func rollbackTransaction() async throws {}
}

File: packages/kits/PersistenceKit/rust/src/row_store.rs
//! RowStore trait: typed row I/O.

use crate::error::StorageResult;
use crate::predicate::{OrderClause, StoragePredicate};
use crate::types::{RowHandle, StorageRow, TypedValue};
use std::collections::BTreeMap;
use substrate_types::AsOfCoordinate;

pub trait RowStore: Send + Sync {
    fn insert(&self, table: &str, values: BTreeMap<String, TypedValue>)
        -> StorageResult<RowHandle>;

    fn upsert(
        &self,
        table: &str,
        values: BTreeMap<String, TypedValue>,
        conflict_columns: &[String],
    ) -> StorageResult<RowHandle>;

    fn update(
        &self,
        table: &str,
        values: BTreeMap<String, TypedValue>,
        predicate: &StoragePredicate,
    ) -> StorageResult<usize>;

    fn delete(&self, table: &str, predicate: &StoragePredicate) -> StorageResult<usize>;

    fn query(
        &self,
        table: &str,
        predicate: Option<&StoragePredicate>,
        order_by: &[OrderClause],
        limit: Option<usize>,
        offset: Option<usize>,
    ) -> StorageResult<Vec<StorageRow>>;

    /// Column-projected query: identical to [`query`](Self::query) but only the
    /// columns named in `columns` appear in each returned [`StorageRow`].
    ///
    /// This is the storage-layer hook for the recall no-blob path: a
    /// `.structured` recall projects away the `content` column so the decoded
    /// drawer carries `content == ""` (LocusKit spec § 7.3) without paying the
    /// blob I/O. It is the Rust parity of the Swift column-projection recall
    /// query.
    ///
    /// The default delegates to [`query`](Self::query) and then drops every
    /// column not in `columns` — correct for any backend, but it still reads
    /// the full row from storage. Backends that can push the projection into
    /// the engine (SQLite, PostgreSQL: `SELECT col1, col2, …`) override this so
    /// the omitted column is never read off disk. An empty `columns` slice is
    /// treated as "no projection" and returns full rows, matching `query`.
    fn query_projected(
        &self,
        table: &str,
        columns: &[&str],
        predicate: Option<&StoragePredicate>,
        order_by: &[OrderClause],
        limit: Option<usize>,
        offset: Option<usize>,
    ) -> StorageResult<Vec<StorageRow>> {
        let rows = self.query(table, predicate, order_by, limit, offset)?;
        if columns.is_empty() {
            return Ok(rows);
        }
        Ok(rows
            .into_iter()
            .map(|row| {
                let mut kept: BTreeMap<String, TypedValue> = BTreeMap::new();
                for &c in columns {
                    if let Some(v) = row.get(c) {
                        kept.insert(c.to_string(), v.clone());
                    }
                }
                StorageRow::new(kept)
            })
            .collect())
    }

    fn count(&self, table: &str, predicate: Option<&StoragePredicate>) -> StorageResult<usize>;

    /// Column-projected variant of [`query_skip_corrupt`](Self::query_skip_corrupt).
    ///
    /// Identical semantics to `query_skip_corrupt` — rows with
    /// `StorageError::CorruptStoredValue` on any column are skipped and logged;
    /// other errors abort — but only the columns named in `columns` are returned
    /// in each row, matching the projection contract of
    /// [`query_projected`](Self::query_projected).
    ///
    /// An empty `columns` slice is treated as "no projection" and returns full
    /// rows (matching `query_projected`'s convention), skipping corrupt rows.
    ///
    /// The default implementation delegates to `query_projected` and wraps a
    /// single top-level `CorruptStoredValue` error as `(vec![], 1)`.
    /// Backends that can implement row-level skipping at the cursor (SQLite)
    /// override this for correct per-row skip-and-log behaviour.
    fn query_projected_skip_corrupt(
        &self,
        table: &str,
        columns: &[&str],
        predicate: Option<&StoragePredicate>,
        order_by: &[OrderClause],
        limit: Option<usize>,
        offset: Option<usize>,
    ) -> StorageResult<(Vec<StorageRow>, usize)> {
        match self.query_projected(table, columns, predicate, order_by, limit, offset) {
            Ok(rows) => Ok((rows, 0)),
            Err(crate::error::StorageError::CorruptStoredValue { table: t, column: c, stored_text: s }) => {
                eprintln!(
                    "[persistence_kit] WARNING: query_projected_skip_corrupt fallback: at least \
                     one corrupt row in '{}' (column='{}' stored_text='{}'). Backend does not \
                     support row-level skipping; returning empty result.",
                    t, c, s
                );
                Ok((vec![], 1))
            }
            Err(other) => Err(other),
        }
    }

    /// Corpus scan that skips rows with a corrupt stored value rather than
    /// failing the entire scan.
    ///
    /// ## When to use
    ///
    /// Use this method for **best-effort corpus scans** (e.g. `all_drawers`,
    /// `drawers_in_wing`) where a single corrupt row must not brick the entire
    /// estate. One bad timestamp or UUID should cause that row to be logged and
    /// skipped, not abort the whole query. For **point lookups** (single-row
    /// fetches by primary key) use [`query`](Self::query), which is strict: a
    /// corrupt value in a point-lookup row is an unambiguous data-integrity
    /// failure and the caller must know about it.
    ///
    /// ## Contract
    ///
    /// - Returns `(clean_rows, skipped_count)`.
    /// - Rows that decode without error appear in `clean_rows`.
    /// - Rows that fail with `StorageError::CorruptStoredValue` are counted in
    ///   `skipped_count` and written to stderr as a warning (the log line names
    ///   the table, column, and stored text so the bad value can be identified).
    /// - Any **other** error (SQL engine failure, connectivity, locking) is
    ///   re-raised as `Err(…)` — those are systemic failures, not data problems.
    ///
    /// ## Default implementation
    ///
    /// Calls [`query`](Self::query) and wraps any `CorruptStoredValue` as
    /// `(vec![], 1)`. Backends that can implement row-level skipping in the
    /// engine loop (SQLite's rusqlite row iterator) override this to avoid
    /// aborting the cursor on a corrupt row.
    fn query_skip_corrupt(
        &self,
        table: &str,
        predicate: Option<&StoragePredicate>,
        order_by: &[OrderClause],
        limit: Option<usize>,
        offset: Option<usize>,
    ) -> StorageResult<(Vec<StorageRow>, usize)> {
        // Default: call strict query; promote CorruptStoredValue to a skip.
        match self.query(table, predicate, order_by, limit, offset) {
            Ok(rows) => Ok((rows, 0)),
            Err(crate::error::StorageError::CorruptStoredValue { table: t, column: c, stored_text: s }) => {
                eprintln!(
                    "[persistence_kit] WARNING: query_skip_corrupt fallback: at least one \
                     corrupt row in '{}' (column='{}' stored_text='{}'). Backend does not \
                     support row-level skipping; returning empty result.",
                    t, c, s
                );
                Ok((vec![], 1))
            }
            Err(other) => Err(other),
        }
    }

    // ----------------------------------------------------------------
    // As-of temporal query (ADR-017 §15)
    // ----------------------------------------------------------------

    /// Temporal query: returns rows visible at the given `AsOfCoordinate`.
    ///
    /// - `None` or `Present`: delegates to [`query`](Self::query) — the
    ///   current live state.
    /// - `AsOf(hlc)`: returns rows whose HLC validity range includes the
    ///   given HLC. **Currently gated off** — returns
    ///   `StorageError::FeatureGated` until NT-L4 (lineage-wide expunge)
    ///   and NT-P3 (erasure overlay) have both merged.
    ///
    /// The filter logic (when ungated): a row with `created_hlc` and optional
    /// `tombstoned_hlc` is visible at HLC T when
    /// `created_hlc <= T AND (tombstoned_hlc IS NULL OR tombstoned_hlc > T)`.
    fn query_as_of(
        &self,
        table: &str,
        predicate: Option<&StoragePredicate>,
        order_by: &[OrderClause],
        limit: Option<usize>,
        offset: Option<usize>,
        as_of: Option<AsOfCoordinate>,
    ) -> StorageResult<Vec<StorageRow>> {
        match as_of {
            None | Some(AsOfCoordinate::Present) => {
                self.query(table, predicate, order_by, limit, offset)
            }
            Some(AsOfCoordinate::AsOf(_)) => {
                Err(crate::error::StorageError::FeatureGated {
                    feature: "asOfQuery".to_owned(),
                })
            }
        }
    }

    /// Temporal projected query: as-of variant of [`query_projected`](Self::query_projected).
    /// Same gating behavior as [`query_as_of`](Self::query_as_of).
    fn query_projected_as_of(
        &self,
        table: &str,
        columns: &[&str],
        predicate: Option<&StoragePredicate>,
        order_by: &[OrderClause],
        limit: Option<usize>,
        offset: Option<usize>,
        as_of: Option<AsOfCoordinate>,
    ) -> StorageResult<Vec<StorageRow>> {
        match as_of {
            None | Some(AsOfCoordinate::Present) => {
                self.query_projected(table, columns, predicate, order_by, limit, offset)
            }
            Some(AsOfCoordinate::AsOf(_)) => {
                Err(crate::error::StorageError::FeatureGated {
                    feature: "asOfQuery".to_owned(),
                })
            }
        }
    }

    /// Temporal skip-corrupt query: as-of variant of
    /// [`query_skip_corrupt`](Self::query_skip_corrupt).
    /// Same gating behavior as [`query_as_of`](Self::query_as_of).
    fn query_skip_corrupt_as_of(
        &self,
        table: &str,
        predicate: Option<&StoragePredicate>,
        order_by: &[OrderClause],
        limit: Option<usize>,
        offset: Option<usize>,
        as_of: Option<AsOfCoordinate>,
    ) -> StorageResult<(Vec<StorageRow>, usize)> {
        match as_of {
            None | Some(AsOfCoordinate::Present) => {
                self.query_skip_corrupt(table, predicate, order_by, limit, offset)
            }
            Some(AsOfCoordinate::AsOf(_)) => {
                Err(crate::error::StorageError::FeatureGated {
                    feature: "asOfQuery".to_owned(),
                })
            }
        }
    }

    // ----------------------------------------------------------------
    // Explicit transaction boundary (GLK_BATCH1)
    // ----------------------------------------------------------------

    /// Open a serializable write transaction on the backing store.
    ///
    /// Backends that support explicit transactions (`SqliteRowStore`) override
    /// this to issue `BEGIN IMMEDIATE`. All other backends inherit the no-op
    /// default, which is correct — they have no serializable multi-statement
    /// transaction concept.
    ///
    /// `CachingRowStore` explicitly delegates to its backing store rather than
    /// relying on the no-op default: live GLK estates wrap `SqliteRowStore` in
    /// a `CachingRowStore`, so the default would silently swallow the boundary.
    fn begin_transaction(&self) -> StorageResult<()> {
        Ok(())
    }

    /// Commit the transaction opened by `begin_transaction`.
    ///
    /// Backends override alongside `begin_transaction`. Same no-op policy.
    fn commit_transaction(&self) -> StorageResult<()> {
        Ok(())
    }

    /// Roll back the transaction opened by `begin_transaction`, discarding
    /// all changes since `BEGIN IMMEDIATE`.
    ///
    /// Backends override alongside `begin_transaction`. Same no-op policy.
    fn rollback_transaction(&self) -> StorageResult<()> {
        Ok(())
    }
}
